Add CommandConverter and CogConverter + add usage in Core

phenom4n4n opened the pull request:
Adds a command and cog converter to the `commands.converter` module, which would be useful in several core commands and various cogs.

Commit:
add commands, cog converter

phenom4n4n committed:
properly use type_checking

phenom4n4n committed:
make core commands use command converter

phenom4n4n committed:
update commands to use cogconverter

phenom4n4n committed:
Merge branch 'V3/develop' into command-converter

phenom4n4n committed:
fix undefined variable name, style

phenom4n4n commented:
This PR is ready for review.

Jackenmen commented on redbot/core/commands/converter.py:
```diff
@@ -416,3 +418,33 @@ def __class_getitem__(cls, k):
+    class CommandConverter(dpy_commands.Converter):
+        """
+        Returns a `Command` object if the passed argument matches a command on the bot.
+        """
```
```suggestion
        """Converts a command name to the matching `Command` object."""
```

Jackenmen commented on redbot/core/commands/converter.py:
```diff
@@ -416,3 +418,33 @@ def __class_getitem__(cls, k):
+    class CogConverter(dpy_commands.Converter):
+        """
+        Returns a `Cog` object if the passed argument matches a cog name on the bot.
+        """
```
```suggestion
        """Converts a cog name to the matching `Cog` object."""
```

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -1151,15 +1154,9 @@ async def embedset_showsettings(self, ctx: commands.Context, command_name: str =
         **Arguments:**
             - `[command_name]` - Checks this command for command specific embed settings.
```
Argument name was changed: 
```suggestion
            - `[command]` - Checks this command for command specific embed settings.
```
This also needs equivalent change in the docs.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -4060,7 +4045,7 @@ async def command_manager(self, ctx: commands.Context):
     @checks.is_owner()
     @command_manager.command(name="defaultdisablecog")
-    async def command_default_disable_cog(self, ctx: commands.Context, *, cogname: str):
+    async def command_default_disable_cog(self, ctx: commands.Context, *, cogname: CogConverter):
```
As with command_name->command, this should also be renamed from cogname to cog:
```suggestion
    async def command_default_disable_cog(self, ctx: commands.Context, *, cog: CogConverter):
```
You will need to update the docstring and docs as well.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -4075,17 +4060,16 @@ async def command_default_disable_cog(self, ctx: commands.Context, *, cogname: s
     @checks.is_owner()
     @command_manager.command(name="defaultenablecog")
-    async def command_default_enable_cog(self, ctx: commands.Context, *, cogname: str):
+    async def command_default_enable_cog(self, ctx: commands.Context, *, cogname: CogConverter):
```
As with command_name->command, this should also be renamed from cogname to cog:
```suggestion
    async def command_default_enable_cog(self, ctx: commands.Context, *, cog: CogConverter):
```
You will need to update the docstring and docs as well.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -4100,15 +4084,14 @@ async def command_default_enable_cog(self, ctx: commands.Context, *, cogname: st
     @commands.guild_only()
     @command_manager.command(name="disablecog")
-    async def command_disable_cog(self, ctx: commands.Context, *, cogname: str):
+    async def command_disable_cog(self, ctx: commands.Context, *, cogname: CogConverter):
```
As with command_name->command, this should also be renamed from cogname to cog:
```suggestion
    async def command_disable_cog(self, ctx: commands.Context, *, cog: CogConverter):
```
You will need to update the docstring and docs as well.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -4134,7 +4116,7 @@ async def command_disable_cog(self, ctx: commands.Context, *, cogname: str):
     @commands.guild_only()
     @command_manager.command(name="enablecog")
-    async def command_enable_cog(self, ctx: commands.Context, *, cogname: str):
+    async def command_enable_cog(self, ctx: commands.Context, *, cogname: CogConverter):
```
As with command_name->command, this should also be renamed from cogname to cog:
```suggestion
    async def command_enable_cog(self, ctx: commands.Context, *, cog: CogConverter):
```
You will need to update the docstring and docs as well.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -4276,39 +4260,32 @@ async def command_disable_global(self, ctx: commands.Context, *, command: str):
             return
 
         async with ctx.bot._config.disabled_commands() as disabled_commands:
             if command not in disabled_commands:
```
`command` used to be a string so this needs to be updated appropriately:
```suggestion
            if command.qualified_name not in disabled_commands:
```

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -4319,42 +4296,35 @@ async def command_disable_guild(self, ctx: commands.Context, *, command: str):
             return
 
         async with ctx.bot._config.guild(ctx.guild).disabled_commands() as disabled_commands:
             if command not in disabled_commands:
```
`command` used to be a string so this needs to be updated appropriately:
```suggestion
            if command.qualified_name not in disabled_commands:
```

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -1249,7 +1246,7 @@ async def embedset_guild(self, ctx: commands.Context, enabled: bool = None):
     @embedset.group(name="command", invoke_without_command=True)
     async def embedset_command(
-        self, ctx: commands.Context, command_name: str, enabled: bool = None
+        self, ctx: commands.Context, command: CommandConverter, enabled: bool = None
```
The argument name was changed so this also needs equivalent change in the docs.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -1289,7 +1286,7 @@ def _check_if_command_requires_embed_links(self, command_obj: commands.Command)
     @embedset_command.command(name="global")
     async def embedset_command_global(
-        self, ctx: commands.Context, command_name: str, enabled: bool = None
+        self, ctx: commands.Context, command: CommandConverter, enabled: bool = None
```
The argument name was changed so this also needs equivalent change in the docs.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -1340,7 +1331,7 @@ async def embedset_command_global(
     @embedset_command.command(name="server", aliases=["guild"])
     async def embedset_command_guild(
-        self, ctx: commands.GuildContext, command_name: str, enabled: bool = None
+        self, ctx: commands.GuildContext, command: CommandConverter, enabled: bool = None
```
The argument name was changed so this also needs equivalent change in the docs.

phenom4n4n committed:
Update redbot/core/commands/converter.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/core/commands/converter.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/core/core_commands.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/core/core_commands.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/core/core_commands.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/core/core_commands.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
Update redbot/core/core_commands.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

phenom4n4n committed:
update cog argument names

Commit:
update documentation arg names

Commit:
update more docs

Jackenmen committed:
Merge branch 'V3/develop' into command-converter

Jackenmen committed:
This new Sphinx is annoying about this...

Jackenmen committed:
I'm questioning my skills

Jackenmen committed:
Fix name error in `[p]embedset showsettings` when command is not given

Jackenmen committed:
Do not use the new cog converter in `[p]command enablecog`

This is needed so that a cog that isn't loaded but was disabled
can be enabled back.

Jackenmen reviewed (approved):
Looks good to me, thanks for the PR!

![](https://i.fluffy.cc/27MHQpGVHf8zsBqx96gxT5PB5mS1jtjD.gif)
